app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -0,0 +1,379 @@
+      const stageData = stageResults.stages[stageName];
+      if (stageData.job_status !== "STARTED") {
+        graph.afterDrawingOnce(() => {
+          this.toggleStage(i);
+        });
+      }
+    });
+  }
+
+  drawStageGraph(index) {
+    const { nodeColor, backgroundColor, edgeColor } = this.props;
+    const container = this[`container-${index}`];
+
+    const edgeData = this.generateIntraEdgeData(index).concat(
+      this.generateInterEdgeData(index)
+    );
+    const nodeData = this.generateNodeData(index, edgeData);
+
+    const options = {
+      nodes: {
+        borderWidth: 0,
+        color: nodeColor,
+        shape: "box",
+        shapeProperties: {
+          borderRadius: 6,
+        },
+        widthConstraint: {
+          minimum: 120,
+        },
+        heightConstraint: {
+          minimum: 24,
+        },
+        font: {
+          face: "Open Sans",
+        },
+      },
+      groups: {
+        startEndNodes: {
+          widthConstraint: 8,
+          heightConstraint: 0,
+          color: backgroundColor,
+          fixed: {
+            x: true,
+            y: true,
+          },
+        },
+      },
+      edges: {
+        arrows: {
+          to: {
+            enabled: true,
+            type: "arrow",
+            scaleFactor: 0.8,
+          },
+        },
+        smooth: {
+          type: "cubicBezier",
+          roundness: 0.8,
+        },
+        color: edgeColor,
+      },
+      layout: {
+        hierarchical: {
+          direction: "LR",
+          sortMethod: "directed",
+          levelSeparation: 200,
+          parentCentralization: false,
+          blockShifting: false,
+          edgeMinimization: false,
+        },
+      },
+      physics: {
+        enabled: false,
+      },
+      interaction: {
+        zoomView: false,
+        dragView: false,
+        dragNodes: false,
+      },
+    };
+
+    const currStageGraph = new NetworkGraph(
+      container,
+      nodeData,
+      edgeData,
+      options
+    );
+    currStageGraph.minimizeWidthGivenScale(1.0);
+
+    this.graphs.push(currStageGraph);
+  }
+
+  render() {
+    const stageContainers = [];
+
+    this.stageNames.forEach((stageName, i) => {
+      const isOpened = this.state.stagesOpened[i];
+
+      stageContainers.push(
+        <div key={stageName} className={cs.stage}>
+          <div
+            className={isOpened ? cs.hidden : cs.stageButton}
+            onClick={() => this.toggleStage(i)}
+          >
+            {stageName}
+          </div>
+
+          <div className={isOpened ? cs.openedStage : cs.hidden}>
+            <div className={cs.graphLabel}>
+              {stageName}
+              <RemoveIcon
+                className={cs.closeButton}
+                onClick={() => this.toggleStage(i)}
+              />
+            </div>
+            <div
+              className={cs.graph}
+              ref={ref => {
+                this[`container-${i}`] = ref;
```

Comment:
you could make an array `this.graphContainers[i] = ref` and initialize `this.graphContainers = []` in the constructor.